### DISCUSSION TITLE
Files getting renamed by onedrive

### DISCUSSION THREAD

**Discussion by tom-andersson** [abraunegg/onedrive#1765] [Q&A] [answered]
I've noticed a bunch of files tracked by `onedrive` being renamed with an extended filename. The files are renamed like `filename-<name-of-machine>.extension`. I went into the logs from `journalctl` to search for the point this occurred for one example. Below is the output I found:
```
Dec 21 15:22:45 bsl-tomand-l1 onedrive[5833]: Uploading modified file vimwiki/diary/2021-12-14.wiki ... skipped.
Dec 21 15:22:46 bsl-tomand-l1 onedrive[5833]: Uploading new file vimwiki/diary/2021-12-14.wiki ... Cannot upload file changes/creation: Cannot open file `vimwiki/diary/2021-12-14.wiki' in mode `rb' (No such file or directory)
Dec 21 15:22:46 bsl-tomand-l1 onedrive[5833]: Moving vimwiki/diary/2021-12-14.wiki to vimwiki/diary/2021-12-14-bsl-tomand-l1.wiki
Dec 21 15:22:46 bsl-tomand-l1 onedrive[5833]: Uploading new file vimwiki/diary/2021-12-14-bsl-tomand-l1.wiki ...
Dec 21 15:22:48 bsl-tomand-l1 onedrive[5833]: [470B blob data]
Dec 21 15:22:48 bsl-tomand-l1 onedrive[5833]: done.
Dec 21 15:22:48 bsl-tomand-l1 onedrive[5833]: Cannot upload file changes/creation: vimwiki/diary/2021-12-14.wiki: No such file or directory
```
My naive understanding is that:
* the client attempts to upload the modified file (which also exists server-side), but it is skipped for some reason
* It then tries to upload the file as if it were a new file, but there is a clash with the existing file on the server
* The file gets renamed locally and then uploaded

No idea if the above is the correct series of events. Would appreciate an answer as to why this is happening, or knowing if the logs above aren't enough to pinpoint the issue.

**Comment by abraunegg** [comment, marked as answer]
@tom-andersson 
> No idea if the above is the correct series of events. Would appreciate an answer as to why this is happening, or knowing if the logs above aren't enough to pinpoint the issue.

What client version are you using?

Clashes occur when the client cannot determine the correct state of the file from OneDrive.

Typically this occurs when the file is newer locally than remote, but the user keeps running the client using `--resync` .... (do not use this option unless the client advises to use it, it is a one time use option).

When using `--resync` it removes what the client knows about your files - thus, does not know what file is what - hence, to avoid data loss, renames files.

If you are constantly seeing this without using `--resync` when starting the client, use `--local-first` to tell the client your local data should be considered the 'source-of-truth'

>```Dec 21 15:22:45 bsl-tomand-l1 onedrive[5833]: Uploading modified file vimwiki/diary/2021-12-14.wiki ... skipped.
>Dec 21 15:22:46 bsl-tomand-l1 onedrive[5833]: Uploading new file vimwiki/diary/2021-12-14.wiki ... Cannot upload file >changes/creation: Cannot open file `vimwiki/diary/2021-12-14.wiki' in mode `rb' (No such file or directory)
>Dec 21 15:22:46 bsl-tomand-l1 onedrive[5833]: Moving vimwiki/diary/2021-12-14.wiki to vimwiki/diary/2021-12-14-bsl-tomand-l1.wiki
>Dec 21 15:22:46 bsl-tomand-l1 onedrive[5833]: Uploading new file vimwiki/diary/2021-12-14-bsl-tomand-l1.wiki ...
>```

Looking at this a but deeper as well - the second line is rather telling - cannot open the file.

This is probably a user|group file and directory permissions issue. You are potentially running the client as one user, but then editing the files as a different user - thus - the client cannot access the files, but it can move the files to the renamed one as the client has permission to do so.

Can you please check your file|folder permissions on your data.

**Reply by tom-andersson** [reply]
Thank you for the answer @abraunegg.

- My `onedrive --version` is `v2.4.14-6-ga2d2ea9`.
- It's possible that all the renamed files appeared from a time when I was running `resync` from two separate machines quite often. I will continue to monitor my local data to see if they continue to appear. If so, I'll look into and try the`local-first` option.
- Regarding permissions, I have looked at the permissions of a few of my files using `ls -la`:
```
-rw-------  1 tomand tomand  1497 Dec 20 13:13  2021-12-10.wiki
-rw-rw-r--  1 tomand tomand  3183 Dec 30 17:43  2021-12-14.wiki
-rw-------  1 tomand tomand  1359 Dec 21 00:07  2021-12-20.wiki
-rw-rw-r--  1 tomand tomand   302 Dec 30 11:05  2021-12-22.wiki
```
Most of my files only have read-write access for my main `tomand` user, but some give read permissions to 'group' and 'other'. Not sure why.

**Reply by tom-andersson** [reply]
Hi @abraunegg, unfortunately I'm getting more renamed files. Here's my setup:
* `systemctl` onedrive service running on Laptop 1
* modify a file on that laptop
* `journalctl` outputs `Uploading modified file <path>... done`; `Deleting item from OneDrive <path>`; `Starting a sync with OneDrive`;
* when the `monitor_interval` is elapsed there is the output: `Syncing changes from OneDrive...`; `Uploading new file <path>`
* on Laptop 2, run `onedrive --synchronize --download-only --verbose`
* get the following output:
```
...
The local item has a different modified time 2022-Jan-05 12:25:53Z when compared to remote modified time 2022-Jan-05 12:28:41Z
The local item has a different hash when compared to remote item hash
Remote item modified time is newer based on UTC time conversion
The local item is out-of-sync with OneDrive, renaming to preserve existing file and prevent data loss: vimwiki/unix/Git.wiki -> vimwiki/unix/Git-tomand-hp-envy.wiki
Downloading file vimwiki/unix/Git.wiki ... done.
Trying to delete item vimwiki/unix/Git.wiki
Skipped due to id difference!
```

I'm using this package for syncing my personal wiki between laptops. Having the filesystem littered with various versions of the same file is not ideal, and can result in wiki links pointing to older versions of files. So I would greatly appreciate a solution to this.

I'm confused about a couple things in the logs that may be relevant:
* Why is the `systemctl` service on Laptop 1 doing `Uploading modified file... Deleting item...` followed by `Uploading new file...` at the next sync point? Why not a single upload to the server? It's as if the `onedrive` client can't tell it's the same file on the server, leading to it being overwritten and the state of that file being reset.
* Why would the hashes of the remote and local files on Laptop 2 be different? Is this the same as the 'id difference' in the last line?

**Comment by abraunegg** [comment]
@tom-andersson 
What account type are you using for OneDrive - Personal Account, Business Account or a SharePoint Site?

Why the application does this is because of Business | SharePoint services *modify* your files post upload, by adding Microsoft XML data to the file. Run the client manually without any systemd service running in verbose mode (`--verbose`) and you can see additional details as what and why this is occurring.

> Why would the hashes of the remote and local files on Laptop 2 be different? Is this the same as the 'id difference' in the last line?

Because most likely reason is the above point. Run the client manually on both laptops to work through what is going on.

The application also has the following configuration setting in the config file only:
```
bypass_data_preservation = "false"
```

Set this to "true" and restart your clients.

With this set, when there is a local conflict due to whatever reason, the local file will not be renamed. Use with caution - this may lead to data loss if you are not careful.
